### PR TITLE
Subscription send only to others 

### PR DESCRIPTION
**Pull request by Stevemoretz**
- [x] Added or updated tests
- [X] Documented user facing changes
- [X] Updated CHANGELOG.md

A fix for https://github.com/nuwave/lighthouse/issues/2293

**Changes**

Subscriptions can now be filtered via `$subscriber->socket_id` and `request()->header("x-socket-id")`, which is similar to what https://laravel.com/docs/9.x/broadcasting#only-to-others does.

**Breaking changes**

Nope.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ You can find and compare releases at the [GitHub release page](https://github.co
 - Require implementations of `BatchedEntityResolver` to maintain the keys given in `array $representations` https://github.com/nuwave/lighthouse/pull/2286
 - Use the strongest possible property types over PHPDocs
 - Require filter directives such as `@whereKey` in `@delete`, `@forceDelete` and `@restore` https://github.com/nuwave/lighthouse/pull/2289
+- Subscriptions can now be filtered via `$subscriber->socket_id` and `request()->header("x-socket-id")` https://github.com/nuwave/lighthouse/pull/2298
```
Comment (spawnia) [inline CHANGELOG.md:21]:
```suggestion
- Subscriptions can now be filtered via `$subscriber->socket_id` and `request()->header('X-Socket-ID')` https://github.com/nuwave/lighthouse/pull/2298
```

File: docs/5/subscriptions/filtering-subscriptions.md [hunk 2]
```diff
@@ -40,3 +40,38 @@ class PostUpdatedSubscription extends GraphQLSubscription
     }
 }
 ```
+
+## Only To Others
+
+When building an application that utilizes event broadcasting, you may occasionally need to broadcast an event to all subscribers to a given channel except for the current user. You may accomplish this using the filter function, this following snippet is the equivalent of [toOthers method from Laravel's broadcast helper](https://laravel.com/docs/9.x/broadcasting#only-to-others).
```
Comment (spawnia) [inline docs/5/subscriptions/filtering-subscriptions.md:46]:
```suggestion
When building an application that utilizes event broadcasting, you may occasionally need to broadcast an event to all subscribers of a channel except for the current user.
You may accomplish this using the filter function, this following snippet is equivalent to [the `toOthers()` method from Laravel's broadcast helper](https://laravel.com/docs/9.x/broadcasting#only-to-others).
```

File: docs/5/subscriptions/filtering-subscriptions.md [hunk 3]
```diff
@@ -40,3 +40,38 @@ class PostUpdatedSubscription extends GraphQLSubscription
...
+
+```php
+namespace App\GraphQL\Subscriptions;
+
+use Nuwave\Lighthouse\Subscriptions\Subscriber;
+use Nuwave\Lighthouse\Schema\Types\GraphQLSubscription;
+
+class PostUpdatedSubscription extends GraphQLSubscription
```
Comment (spawnia) [inline docs/5/subscriptions/filtering-subscriptions.md:54]:
```suggestion
final class PostUpdatedSubscription extends GraphQLSubscription
```

File: docs/5/subscriptions/filtering-subscriptions.md [hunk 4]
```diff
@@ -40,3 +40,38 @@ class PostUpdatedSubscription extends GraphQLSubscription
...
+     * Filter which subscribers should receive the subscription.
+     *
+     * @param  \Nuwave\Lighthouse\Subscriptions\Subscriber  $subscriber
+     * @param  mixed  $root
+     * @return bool
+     */
+    public function filter(Subscriber $subscriber, $root): bool
+    {
+        // Filter out the sender
+        return $subscriber->socket_id !== request()->header("x-socket-id");
```
Comment (spawnia) [inline docs/5/subscriptions/filtering-subscriptions.md:66]:
```suggestion
        return $subscriber->socket_id !== request()->header('X-Socket-ID');
```

File: docs/5/subscriptions/filtering-subscriptions.md [hunk 5]
```diff
@@ -40,3 +40,38 @@ class PostUpdatedSubscription extends GraphQLSubscription
...
+class PostUpdatedSubscription extends GraphQLSubscription
+{
+    /**
+     * Filter which subscribers should receive the subscription.
+     *
+     * @param  \Nuwave\Lighthouse\Subscriptions\Subscriber  $subscriber
+     * @param  mixed  $root
+     * @return bool
+     */
+    public function filter(Subscriber $subscriber, $root): bool
```
Comment (spawnia) [inline docs/5/subscriptions/filtering-subscriptions.md:63]:
```suggestion
     */
    public function filter(Subscriber $subscriber, mixed $root): bool
```

See https://github.com/nuwave/lighthouse/pull/2299

File: docs/5/subscriptions/filtering-subscriptions.md [hunk 6]
```diff
@@ -40,3 +40,38 @@ class PostUpdatedSubscription extends GraphQLSubscription
...
+    {
+        // Filter out the sender
+        return $subscriber->socket_id !== request()->header("x-socket-id");
+    }
+}
+```
+
+When you initialize a Laravel Echo instance, a socket ID is assigned to the connection. If you are using a global [Axios](https://github.com/mzabriskie/axios) instance to make HTTP requests from your JavaScript application, the socket ID will automatically be attached to every outgoing request as a `X-Socket-ID` header. Then, you can access that in your filter function.
```
Comment (spawnia) [inline docs/5/subscriptions/filtering-subscriptions.md:71]:
```suggestion
When you initialize a Laravel Echo instance, a socket ID is assigned to the connection.
If you are using a global [Axios](https://github.com/mzabriskie/axios) instance to make HTTP requests from your JavaScript application, the socket ID will automatically be attached to every outgoing request in the `X-Socket-ID` header.
Then, you can access that in your filter function.
```

File: docs/5/subscriptions/filtering-subscriptions.md [hunk 7]
```diff
@@ -40,3 +40,38 @@ class PostUpdatedSubscription extends GraphQLSubscription
...
+
+If you are not using a global Axios instance, you will need to manually configure your JavaScript application to send the X-Socket-ID header with all outgoing requests. You may retrieve the socket ID using the Echo.socketId method:
```
Comment (spawnia) [inline docs/5/subscriptions/filtering-subscriptions.md:73]:
```suggestion
If you are not using a global Axios instance, you will need to manually configure your JavaScript application to send the `X-Socket-ID` header with all outgoing requests.
You may retrieve the socket ID using the `Echo.socketId()` method:
```

File: docs/5/subscriptions/filtering-subscriptions.md [hunk 8]
```diff
@@ -40,3 +40,38 @@ class PostUpdatedSubscription extends GraphQLSubscription
...
+
+```js
+var socketId = Echo.socketId();
```
Comment (spawnia) [inline docs/5/subscriptions/filtering-subscriptions.md:76]:
```suggestion
const socketId = Echo.socketId();
```

File: src/Subscriptions/Subscriber.php [hunk 9]
```diff
@@ -24,6 +24,11 @@ class Subscriber
      */
     public $channel;
 
+    /**
+     * X-SOCKET-ID header passed on the subscription query.
```
Comment (spawnia) [inline src/Subscriptions/Subscriber.php:28]:
```suggestion
     * X-Socket-ID header passed on the subscription query.
```

File: src/Subscriptions/Subscriber.php [hunk 10]
```diff
@@ -90,6 +95,7 @@ public function __construct(
         $this->args = $args;
         $this->variables = $resolveInfo->variableValues;
         $this->context = $context;
+        $this->socket_id = request()->header('x-socket-id');
```
Comment (spawnia) [inline src/Subscriptions/Subscriber.php:98]:
```suggestion
        $this->socket_id = request()->header('X-Socket-ID');
```

**Review by spawnia — COMMENTED** [review]
Can you add a test?

Also, please update `docs/master` and do not change `docs/5`

File: docs/master/subscriptions/filtering-subscriptions.md [hunk 11]
```diff
@@ -4,11 +4,11 @@ There are times when you'll need to filter out specific events based on the argu
 
 ```graphql
 subscription onPostUpdated($post_id: ID!) {
-  postUpdated(post_id: $post_id) {
-    id
-    title
-    content
-  }
+    postUpdated(post_id: $post_id) {
+        id
```
Comment (spawnia) [inline docs/master/subscriptions/filtering-subscriptions.md:8]: revert

**Comment by spawnia** [comment]
Not sure if there are any tests for it. Try adding something in https://github.com/nuwave/lighthouse/tree/master/tests/Integration/Subscriptions.

**Comment by Stevemoretz** [comment]
> Not sure if there are any tests for it. Try adding something in https://github.com/nuwave/lighthouse/tree/master/tests/Integration/Subscriptions.

Thanks that's ok, I'll write one for storage of socket_id that should be good enough.

**Comment by Stevemoretz** [comment]
> Not sure if there are any tests for it. Try adding something in https://github.com/nuwave/lighthouse/tree/master/tests/Integration/Subscriptions.

All done test was added too, tested locally all passed, did cs-fixer and php stan locally no errors.

The errors in the CI seem to be unrelated to the commits you might wanna check on them.

**Comment by Stevemoretz** [comment]
> https://github.com/nuwave/lighthouse/actions/runs/4175774729/jobs/7231159976 is definitely related.

Sure you're right, I don't get that locally different php version or dependencies maybe 

```php
/**
     * X-Socket-ID header passed on the subscription query.
     */
    public string|array|null $socket_id;
```

This would fix it for php 8 I think? But it will break it for php 7, what do you suggest? Would this fix it for all versions?

```php
    /**
     * X-Socket-ID header passed on the subscription query.
     * @var string|null|array
     */
    public $socket_id;
```

I'm not really good with different php versions yet I usually use one

File: src/Subscriptions/Subscriber.php [hunk 12]
```diff
@@ -24,6 +24,12 @@ class Subscriber
      */
     public $channel;
 
+    /**
+     * X-Socket-ID header passed on the subscription query.
+     * @var string|null|array<mixed>
```
Comment (spawnia) [inline src/Subscriptions/Subscriber.php:29]: I think we should validate this is either a string or null

File: tests/Integration/Subscriptions/Storage/RedisStorageManagerTest.php [hunk 13]
```diff
@@ -97,10 +97,26 @@ public function testSubscribersByTopic(): void
         $this->assertContainsOnlyInstancesOf(Subscriber::class, $createdSubscribers);
     }
 
+    public function testSocketIDStoredOnSubscribe(): void
+    {
+        /** @var RedisStorageManager $storage */
```
Comment (spawnia) [inline tests/Integration/Subscriptions/Storage/RedisStorageManagerTest.php:102]: use assert

**Review by spawnia — COMMENTED** [review]
good test

File: src/Subscriptions/Subscriber.php (hunk 12 above)
Comment (Stevemoretz) [reply src/Subscriptions/Subscriber.php:29]: Agreed array doesn't really make much sense here yet again if I add this:

```.php
        $XSocketID = request()->header('X-Socket-ID');
        if (! is_array($XSocketID)) {
            $this->socket_id = $XSocketID;
        }
```
I gives an error on my php version (8.1) Laravel (9)

```
 100    Call to function is_array() with string|null will always evaluate to false.
```

Kind of stuck here...

Comment (Stevemoretz) [reply src/Subscriptions/Subscriber.php:29]: And even if I use this:

```php
        $XSocketID = request()->header('X-Socket-ID');
        if (is_string($XSocketID) || is_null($XSocketID)) {
            $this->socket_id = $XSocketID;
        }
```

Php stan gives:

```
  100    Result of || is always true.      
```

Comment (spawnia) [reply src/Subscriptions/Subscriber.php:29]: I would throw if the value is an array. Just try and push it to see if the different versions disagree, then maybe add a `@phpstan-ignore-next-line`. I am fixing the dependencies in https://github.com/nuwave/lighthouse/pull/2300.

**Review by spawnia — APPROVED** [review]

**Comment by spawnia** [comment]
Merging `master` should fix CI.

**Comment by Stevemoretz** [comment]
@spawnia Thanks for working on this with me I enjoy doing PRs with you!

**Comment by spawnia** [comment]
Thanks for the contribution!

**Pull request merged**